### DISCUSSION TITLE
Resources for getting started in research software development

### DISCUSSION THREAD

**Discussion by wkearn** [csdms/project#27] [General]
At the Euro-CSDMS meeting a couple of weeks ago, one of the things we suggested was curating for the CSDMS community some resources on creating and maintaining research software.

What materials do you find particularly useful when working on your own research software projects? What do you wish you had when you started working on research software?

**Comment by mdpiper** [comment]
This may be kinda lame, but over my career as an RSE, the most helpful resource has been web searches for Stack Overflow posts (ignoring all search results that aren't SO posts).

**Comment by gregtucker** [comment]
Over the last several years I have learned a ton by working with the CSDMS Research Software Engineers (RSEs) and seeing how they approach things. I started writing numerical modeling code back around 1991 (!). I wish I had known then about unit testing and test-driven development. Nowadays, I find that when I follow test-driven development, it takes a little bit longer to write the first draft, but a LOT less time to debug it. So I guess my summary is: (1) if you have the opportunity to get feedback/advice from a Research Software Engineer, take advantage of it; (2) take a look at some RSE-curated code as a model; (3) take some time to learn about unit testing and test-driven development (for example, [here](https://csdms.colorado.edu/wiki/Presenters-0478))---it's not that complicated, and can help avoid a lot of headaches down the road!

**Comment by dorchard** [comment]
A couple of resources spring to mind:
* My team put together an RSE skills in Python training set [https://cambridge-iccs.github.io/rse-skills-python/](https://cambridge-iccs.github.io/rse-skills-python/#/title-slide) (it even has a [YouTube video!](https://www.youtube.com/watch?v=nyp9t6mGdsw&list=PL3PByZO-B6dM4pw2AwLepEBsfDrly9L0w&index=8)).
* The 'Good Enough' practice in scientific computing Software Carpentry info: https://swcarpentry.github.io/good-enough-practices-in-scientific-computing/

**Reply by mdpiper** [reply]
On the topic of best practices, Eric and I have been low-key working on a presentation for a target audience of grad students. [Here's the repo](https://github.com/mdpiper/best-practices) and I've hosted it at https://csdms.io/best-practices. If anyone would like to suggest changes (@dorchard has!), I'm very open to them.

**Comment by samharrison7** [comment]
If Python packaging is your thing, then I found this guide really useful when learning how to make Python packages: https://py-pkgs.org/welcome

Me and a few others from the SSI wrote an article a few years back about how to make research software reproducible. It is basically a list of best practices for different "levels" of software (simple tools through to software as infrastructure): https://zenodo.org/records/4761867

**Reply by samharrison7** [reply]
Mark and Eric's version of best practices is much prettier though 😉

**Reply by mdpiper** [reply]
All credit to [reveal-md](https://github.com/webpro/reveal-md)! It's been fun making slides under version control.
